src/execution/vm/llvm_engine.cpp

```diff
@@ -948,29 +951,31 @@ void LLVMEngine::CompiledModuleBuilder::Simplify() {
 }
 
 void LLVMEngine::CompiledModuleBuilder::Optimize() {
-  llvm::legacy::PassManager module_passes;
   llvm::legacy::FunctionPassManager function_passes(llvm_module_.get());
 
-  // Add the appropriate TargetLibraryInfo and TargetTransformInfo
-  auto target_library_info_impl = std::make_unique<llvm::TargetLibraryInfoImpl>(target_machine_->getTargetTriple());
+  // Add the appropriate TargetTransformInfo.
   function_passes.add(llvm::createTargetTransformInfoWrapperPass(target_machine_->getTargetIRAnalysis()));
-  module_passes.add(new llvm::TargetLibraryInfoWrapperPass(*target_library_info_impl));
-  module_passes.add(llvm::createTargetTransformInfoWrapperPass(target_machine_->getTargetIRAnalysis()));
 
-  // Build up optimization pipeline
+  // Build up optimization pipeline.
   llvm::PassManagerBuilder pm_builder;
   pm_builder.OptLevel = 3;
   pm_builder.Inliner = llvm::createFunctionInliningPass(3, 0, false);
   pm_builder.populateFunctionPassManager(function_passes);
-  pm_builder.populateModulePassManager(module_passes);
 
-  // Run optimization passes on module
+  // Add custom passes. Hand-selected based on empirical evaluation.
+  function_passes.add(llvm::createInstructionCombiningPass());
+  function_passes.add(llvm::createReassociatePass());
+  function_passes.add(llvm::createGVNPass());
+  function_passes.add(llvm::createCFGSimplificationPass());
+  function_passes.add(llvm::createAggressiveDCEPass());
+  function_passes.add(llvm::createCFGSimplificationPass());
+
+  // Run optimization passes on all functions.
   function_passes.doInitialization();
-  for (auto &func : *llvm_module_) {
+  for (llvm::Function &func : *llvm_module_) {
```

Comment:
Not a fan of [AAA](https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/)?

Reply:
While ported wholesale from Prashanth's commit, this reminds me of a conversation that I had with another masters student Y a long time ago (who advocated for less use of auto on readability grounds).

I _used_ to agree with that. I now disagree on readability; unless the RHS has an explicit type I will generally prefer not using auto.

I think there is a difference between writing code for experts and writing code for non-experts. Herb Sutter is definitely an expert. But for the average person picking up this project, they will probably be learning LLVM quirks and TPL quirks at the same time while reading this file, and the types are very helpful in learning what exactly is going on. Our interfaces are refactored rarely and having to type out the type doesn't make the interface any better or worse designed.

Reply:
This is actually _not_ one of the C++ style things I'm particularly attached to, and your point about making the LLVM types "more visible" to new people who come to this file convinces me. Also, there is the additional argument to be made that this is closer to "LLVM style" where it seems like they prefer to use explicit types, at least far more often than you might observe in other C++ codebases.